### ISSUE TITLE
IEventRecorder .WithArgs<T> returns unfiltered IEventRecorder - feature / room for improvement?

### ISSUE THREAD

**Issue by joergbattermann** [fluentassertions/fluentassertions#337]
I just noticed an somewhat unexpected behaviour testing events for one of my classes:

Given the following test

```c#
[Fact]
public void AddRangeRaisesPropertyChangedEventForItemIndexer()
{
	// given
	using (var observableDictionary = new ObservableDictionary<int, string>())
	{
		observableDictionary.ThresholdAmountWhenItemChangesAreNotifiedAsReset = 0;
		observableDictionary.MonitorEvents();

		var items = new List<KeyValuePair<int, string>>()
		{
			new KeyValuePair<int, string>(1, "One"),
			new KeyValuePair<int, string>(2, "Two"),
			new KeyValuePair<int, string>(3, "Three"),
		};

		// when
		observableDictionary.AddRange(items);

		// then
		observableDictionary
			.ShouldRaise(nameof(observableDictionary.PropertyChanged))
			.WithSender(observableDictionary)
			.WithArgs<PropertyChangedEventArgs>(args => args.PropertyName == "Item[]")
			.Should().HaveCount(1);
	}
}
```

I get a failing test because Count (the very last line) was expected to be 1, but it is 2.

Now the thing is - exactly one event is being raised for `Item[]`, the second one is for a different property (.Count).

Looking at the source of [PropertyChangedSourceExtensions.cs](https://github.com/dennisdoomen/fluentassertions/blob/e240c8c8d02160fd69e22d7590901df429f86bd3/Src/Core/PropertyChangedSourceExtensions.cs#L164) it becomes apparent that currently the IEventRecorder instance is returned back _as-is_ and not a filtered-down version using the supplied predicate.

I was expecting that every .With*() method here checks for the condition and if it passes, it returns the filtered eventRecorder back so one can actually fluently chain up checks.

I.e.:
- `.ShouldRaise(nameof(observableDictionary.PropertyChanged))` checks for any events for the given name, fails if none are raised or if they are, returns a filtered EventRecorder back only containing those events
- `.WithSender(observableDictionary)` then checks whether those ("PropertyChanged") events have the given sender, fails if not OR if they do, return a further filtered EventRecorder back - only containing "PropertyChanged" events back with the given sender
- `.WithArgs<PropertyChangedEventArgs>(args => args.PropertyName == "Item[]")` - same here: takes the (fluently filtered) stream and checks whether the PropertyName is the expected one, fails if not but if they are, again, returns a further filtered eventRecorder back.. and so on and so on. 

Currently it works 'unfiltered' / fluently for apparently all the .With\* event related methods there, but is this something worth considering/changing or am I just doing/expecting it wrong? :)

Cheers,
-J

**Comment by dennisdoomen** [comment]
I agree with your rationale. Seems we didn't think this through enough (or your the first one to notice this in years)..

**Comment by BrunoJuchli** [comment]
@dennisdoomen 
The rationale for the previous behavior may have been to allow verification that multiple events have happend.
This used to work:

    foo
        .ShouldRaise(nameof(observableDictionary.PropertyChanged))
             .WithArgs<string>(args => args == "some event payload")
             .WithArgs<string>(args => args == "other payload");

And now with FA 6.0 has to be re-written to:

    foo.ShouldRaise(nameof(observableDictionary.PropertyChanged))
             .WithArgs<string>(args => args == "some event payload");
    foo.ShouldRaise(nameof(observableDictionary.PropertyChanged))
             .WithArgs<string>(args => args == "other payload");

I think the 6.0 behavior is just as puzzling as the previous behavior, just in a different way. It allows to combine multiple WithArgs or multiple WithSender etc. calls but this just doesn't make sense the way it's implemented now:

    foo.ShouldRaise(..)
         .WithSender(x)
         .WithSender(y).

Compiles, but doesn't make sense. Ok, fair enough, with the sender probably no-one's ever tempted to write that, because the sender is almost guaranteed to be the object which is raising the event in the first place :D
But same goes for `WithArgs`, where if you actually want to combine conditions you'd rather combine the `Predicate` instead of chaining multiple `WithArgs` calls, like so:

    foo.ShouldRaise(nameof(observableDictionary.PropertyChanged))
         .WithArgs<string>(args => 
             args == "some event payload" &&
             args == "other payload");

**Comment by dennisdoomen** [comment]
I don't think I understand your concern. In v6, a single statement represents a single assertion, so 

```
foo
    .ShouldRaise(nameof(observableDictionary.PropertyChanged))
         .WithArgs<string>(args => args == "some event payload")
         .WithArgs<string>(args => args == "other payload");
```

doesn't really make sense to me. If you were excepting two notifications, you need two separate calls.

**Comment by BrunoJuchli** [comment]
@dennisdoomen
Note: No need to waste your time on this, just read on if your interested :D

I was basically just answering your comment from 5 years ago:

> I agree with your rationale. Seems we didn't think this through enough (...)

Before v6 one didn't need two calls for two notification - which worked exactly like *I* expected it :D
IMO both perspectives/expectations make sense.  Maybe your / original-author past self shared my *current* perspective.

When we stumbled onto this breaking change it took took us quite a while to understand why tests started failing that passed before.

Ideally the syntax (code completion...) would make the intent clear.
IMO it's not obvious that the second `.WithArgs` call operates on a resultset filtered by the first `.WithArgs` call.

I's not like I've got a truly great suggestion at the moment on how to make this idiot proof and super-legible, so like I said at the beginning: no need to waste your time on this ;-)

--------------------------------

Just some food for thought, mostly if I myself ever come back to this. Something like this might be an alternative syntax:

monitor
 .Events["someEvent"]
 .Should(x => x
     .AllHaveSender(testee) // asserts on all
     .AllHaveArgsOfType<string> // asserts on all, returns a new type/syntax considering the just specified args type
     // we now work on something like GenericCollectionAssertions<TArgument>:
     .Equal(
          "some event payload",
          "other payload")));
